l337-postgres/src/lib.rs

```diff
@@ -98,22 +109,36 @@ where
             .await
             .map_err(|e| l337::Error::External(e))?;
 
-        let (sender, receiver) = oneshot::channel();
+        let (done_tx, done_rx) = oneshot::channel();
+        let (drop_tx, drop_rx) = oneshot::channel();
         spawn(async move {
             debug!("connect: start connection future");
-            if let Err(_) = connection.await {
-                sender
-                    .send(true)
-                    .unwrap_or_else(|e| panic!("failed to send shutdown notice: {}", e));
+            let connection = connection.fuse();
+            let drop_rx = drop_rx.fuse();
+
+            futures::pin_mut!(connection, drop_rx);
+
+            futures::select! {
+                result = connection => {
+                    if let Err(e) = result {
+                        warn!("future backing postgres future ended with an error: {}", e);
+                    }
+                }
+                _ = drop_rx => { }
             }
+
+            // If this fails to send, the connection object was already dropped and does not need to be notified
+            let _ = done_tx.send(());
+
             info!("connect: connection future ended");
         });
 
         debug!("connect: postgres connection established");
         Ok(AsyncConnection {
             broken: false,
             client,
-            receiver,
+            done_rx,
+            drop_tx: Some(drop_tx),
```

Comment:
What's the motivation behind the drop channel? I see that there's a select on connection / drop_rx. I don't see where the drop happens in such a way as to be helpful here though. Can you explain?

Reply:
The async postgres/redis libraries both rely on a future being executed in the background that is actually responsible for doing the communication with the network. If this future dies, the client will not be able to proceed, and if the client is Dropped, there is no need for this future to keep running.

I added the Drop channels so that we can ensure when a client is dropped, these driver futures are not being polled unnecessarily any more.

Reply:
And that happens in the wrapper `forget` function right? That was the main thing that was unclear to me on first read is how that happened.

Reply:
Yes, that happens when `forget` is called, causing the connection to be expelled from the pool, or when the whole pool is dropped.

Reply:
I added a comment on the Drop impls in redis/postgres driver crates explaining why I use the channels https://github.com/OneSignal/L3-37/pull/51/files#diff-1739126ceb2c73a0e853b93ebea43e71R38